clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -105,9 +124,107 @@ unit_test_cache_lru_four_way()
+    cache_fifo_test_t cache_fifo_test;
+    cache_fifo_test.initialize_cache(/*associativity=*/8, /*line_size=*/32,
+                                     /*total_size=*/1024);
+    const addr_t ADDRESS_A = 0;
+    const addr_t ADDRESS_B = 128;
+    const addr_t ADDRESS_C = 256;
+    const addr_t ADDRESS_D = 384;
+    const addr_t ADDRESS_E = 512;
+    const addr_t ADDRESS_F = 640;
+    const addr_t ADDRESS_G = 768;
+    const addr_t ADDRESS_H = 896;
+    const addr_t ADDRESS_I = 144;
+    const addr_t ADDRESS_J = 150;
+    const addr_t ADDRESS_K = 900;
+    const addr_t ADDRESS_L = 780;
+
+    assert(cache_fifo_test.get_block_index(ADDRESS_A) ==
+           cache_fifo_test.get_block_index(ADDRESS_B));
+    assert(cache_fifo_test.get_block_index(ADDRESS_B) ==
+           cache_fifo_test.get_block_index(ADDRESS_C));
+    assert(cache_fifo_test.get_block_index(ADDRESS_C) ==
+           cache_fifo_test.get_block_index(ADDRESS_D));
+    assert(cache_fifo_test.get_block_index(ADDRESS_D) ==
+           cache_fifo_test.get_block_index(ADDRESS_E));
+    assert(cache_fifo_test.get_block_index(ADDRESS_F) ==
+           cache_fifo_test.get_block_index(ADDRESS_G));
+    assert(cache_fifo_test.get_block_index(ADDRESS_H) ==
+           cache_fifo_test.get_block_index(ADDRESS_I));
+    assert(cache_fifo_test.get_block_index(ADDRESS_I) ==
+           cache_fifo_test.get_block_index(ADDRESS_J));
+    assert(cache_fifo_test.get_block_index(ADDRESS_J) ==
+           cache_fifo_test.get_block_index(ADDRESS_K));
+    assert(cache_fifo_test.get_block_index(ADDRESS_K) ==
+           cache_fifo_test.get_block_index(ADDRESS_L));
+
+    // Lower-case letter shows the way that is to be replaced after the access
+    // (aka 'first way').
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); // A  x  X  X  X  X  X  X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_B, 2); // A  B  x  X  X  X  X  X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_C, 3); // A  B  C  x  X  X  X  X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_D, 4); // A  B  C  D  x  X  X  X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_E, 5); // A  B  C  D  E  x  X  X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_F, 6); // A  B  C  D  E  F  x  X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_G, 7); // A  B  C  D  F  F  G  x
+    cache_fifo_test.access_and_check_fifo(ADDRESS_H, 0); // a  B  C  D  E  F  G  H
+    cache_fifo_test.access_and_check_fifo(ADDRESS_E, 0); // a  B  C  D  E  F  G  H
+    cache_fifo_test.access_and_check_fifo(ADDRESS_E, 0); // a  B  C  D  E  F  G  H
+    cache_fifo_test.access_and_check_fifo(ADDRESS_E, 0); // a  B  C  D  E  F  G  H
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 0); // a  B  C  D  E  F  G  H
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 0); // a  B  C  D  E  F  G  H
+    cache_fifo_test.access_and_check_fifo(ADDRESS_I, 0); // i  B  C  D  E  F  G  H
+    cache_fifo_test.access_and_check_fifo(ADDRESS_J, 0); // j  B  C  D  E  F  G  H
```

Comment:
Incorrect again: for the same reason.  Ditto below.

Reply:
Done.